src/NetTopologySuite/Geometries/LineString.cs

```diff
@@ -459,39 +459,47 @@ public override void Normalize()
         /// <inheritdoc cref="Geometry.IsEquivalentClass"/>
         protected override bool IsEquivalentClass(Geometry other)
         {
-            return other is LineString;
+            return other is ICurve;
         }
 
         /// <inheritdoc cref="Geometry.CompareToSameClass(object)"/>
         protected internal override int CompareToSameClass(object o)
         {
-            Assert.IsTrue(o is LineString);
-
-            var line = (LineString)o;
-            // MD - optimized implementation
-            int i = 0;
-            int j = 0;
-            while (i < _points.Count && j < line.CoordinateSequence.Count)
+            Assert.IsTrue(o is ICurve);
+            if (o is LineString line)
             {
-                int comparison = _points.GetCoordinate(i).CompareTo(line.CoordinateSequence.GetCoordinate(j));
-                if (comparison != 0)
-                    return comparison;
-                i++;
-                j++;
+                // MD - optimized implementation
+                int i = 0;
+                int j = 0;
+                while (i < _points.Count && j < line.CoordinateSequence.Count)
+                {
+                    int comparison = _points.GetCoordinate(i).CompareTo(line.CoordinateSequence.GetCoordinate(j));
+                    if (comparison != 0)
+                        return comparison;
+                    i++;
+                    j++;
+                }
+
+                if (i < _points.Count)
+                    return 1;
+                if (j < line.CoordinateSequence.Count)
+                    return -1;
+                return 0;
             }
-            if (i < _points.Count)
-                return 1;
-            if (j < line.CoordinateSequence.Count)
-                return -1;
-            return 0;
+
+            return CompareToSameClass(((ICurve) o).Flatten());
         }
 
+        LineString ICurve.Flatten() => this;
+
         /// <inheritdoc cref="Geometry.CompareToSameClass(object, IComparer{CoordinateSequence})"/>
         protected internal override int CompareToSameClass(object o, IComparer<CoordinateSequence> comp)
         {
-            Assert.IsTrue(o is LineString);
-            var line = (LineString)o;
-            return comp.Compare(_points, line.CoordinateSequence);
+            Assert.IsTrue(o is ICurve);
+            if (o is LineString line)
+                return comp.Compare(_points, line.CoordinateSequence);
+
+            return CompareToSameClass(((ICurve) o).Flatten(), comp);
```

Comment:
By looking at `Geometry.CompareTo(object[, IComparer(CoordinateSequence)]` I got the impression that all geometries with the same `SortIndex` should be comparable.

Reply:
> I got the impression that all geometries with the same `SortIndex` should be comparable.

Yes, but only because `SortIndex` should be unique per recognized subclass.  The fact that `SortIndexValue` is defined in the core project is another reason why I claimed in my review summary (and elsewhere) that NetTopologySuite is not currently in a position where it makes sense to try to "plug in" new `Geometry` subclasses externally.

Looking at the details again, even `LineString` and `LinearRing` have different values (I had misremembered when I made my previous comment in this thread), and they're **far** more similar than `LineString` and `CircularString`.